src/compiler/chainrules.jl

```diff
@@ -49,13 +49,13 @@ Convert `x` from the differentials types ChainRules uses  to the format Zygote u
 @inline wrap_chainrules_output(x) = conj(unthunk(x))  # For now we are just not going to deal with thunks
 @inline wrap_chainrules_output(x::Tuple) = map(wrap_chainrules_output, x)
 @inline wrap_chainrules_output(x::ChainRules.AbstractZero) = nothing
-@inline function wrap_chainrules_output(x::ChainRules.Composite{P, T}) where {P, T}
-  T_outer = T <: NamedTuple  ? NamedTuple : Tuple  # must be a Tuple or NamedTuple, don't care about exact parameter types
-  xp = map(wrap_chainrules_output, x)
-  convert(T_outer, xp)
+for T_outer in (:Tuple, :NamedTuple)
+  @eval @inline function wrap_chainrules_output(x::ChainRules.Composite{P, T}) where {P, T<:$T_outer}
+    xp = map(wrap_chainrules_output, x)
+    convert($T_outer, xp)
+  end
```

Comment:
Does this affect the compiler's ability to infer in nested AD?

Reply:
Not entirely sure, but this does remove a branch from the code which the nest AD hates.
I think it also is what lets me (and require me) delete the custom adjoint for Composite

Reply:
Hmm interesting. I was just wondering whether it's possible to write a decent regression test, but possibly not. A comment explaining why this is better done this way rather than the old way would help though.

Reply:
Ok, is this good to go now?

Reply:
Resolved modulo slight rephrasing.